File: packages/core/src/games/7-wonders/rules.ts
import { hasBuiltStageEffect, tableauNames } from "./board";
import { getCardDef } from "./cards";
import { getEdificeDef } from "./edifice";
import { solvePayments } from "./payment";
import type { BuildWonderAction, CardId, GameState, Payment, SevenWondersAction } from "./types";
import { cardIdName } from "./types";
import { getWonderDef } from "./wonders";

/** -1 while everyone selects simultaneously; the pending player's index otherwise. */
export function getActivePlayer(state: GameState): number {
  if (state.phase === "pending" && state.pendingQueue.length > 0) {
    return state.pendingQueue[0].playerIndex;
  }
  return -1;
}

type BuildAction =
  | { type: "play-card"; cardId: CardId; payment: Payment }
  | BuildWonderAction
  | { type: "discard"; cardId: CardId };

/** Every way `playerIndex` could resolve one card: play (chain/pay/free), wonder stage, discard. */
function buildActionsForCard(state: GameState, playerIndex: number, cardId: CardId): BuildAction[] {
  const player = state.players[playerIndex];
  const def = getCardDef(cardIdName(cardId));
  const names = tableauNames(player);
  const actions: BuildAction[] = [{ type: "discard", cardId }];

  // Play the card (never a duplicate name).
  if (!names.has(def.name)) {
    const chain = def.chainFrom?.some((from) => names.has(from)) ?? false;
    if (chain) {
      actions.push({ type: "play-card", cardId, payment: { kind: "chain" } });
    } else {
      const payments = solvePayments(state, playerIndex, def.cost);
      for (const payment of payments ?? []) {
        actions.push({ type: "play-card", cardId, payment });
      }
      if (hasBuiltStageEffect(player, "free-build-per-age") && !player.freeBuildUsedThisAge) {
        actions.push({ type: "play-card", cardId, payment: { kind: "free-build" } });
      }
    }
  }

  // Bury the card to build the next wonder stage (chains/free builds don't apply).
  const side = getWonderDef(player.wonderId).sides[player.side];
  if (player.stagesBuilt < side.stages.length) {
    const stageCost = side.stages[player.stagesBuilt].cost;
    const payments = solvePayments(state, playerIndex, stageCost);

    // Edifice: may also participate in the current Age's project when it's an
    // open project, this player hasn't joined it yet, and they can afford the
    // stage payment plus the participation coin cost.
    const edifice = state.edifices?.[state.age - 1];
    const canParticipate =
      edifice !== undefined &&
      edifice.status === "project" &&
      !edifice.participants.includes(playerIndex);
    const participationCost = canParticipate ? getEdificeDef(edifice.card).cost : 0;

    const stageCoinCost = stageCost.coins ?? 0;
    for (const payment of payments ?? []) {
      if (payment.kind !== "resources") continue;
      actions.push({ type: "build-wonder", cardId, payment });
      const coinsLeft = player.coins - payment.left - payment.right - stageCoinCost;
      if (canParticipate && coinsLeft >= participationCost) {
        actions.push({ type: "build-wonder", cardId, payment, participate: true });
      }
    }
  }

  return actions;
}

export function getLegalActions(state: GameState, playerIndex: number): SevenWondersAction[] {
  if (state.phase === "selecting") {
    if (state.selections[playerIndex] !== null) return [];
    return state.hands[playerIndex].flatMap((cardId) =>
      buildActionsForCard(state, playerIndex, cardId),
    );
  }

  if (state.phase === "pending" && state.pendingQueue[0]?.playerIndex === playerIndex) {
    const pending = state.pendingQueue[0];
    if (pending.kind === "halikarnassos") {
      const names = tableauNames(state.players[playerIndex]);
      const actions: SevenWondersAction[] = [{ type: "skip-pending" }];
      const seen = new Set<string>();
      for (const cardId of state.discard) {
        const name = cardIdName(cardId);
        if (names.has(name) || seen.has(name)) continue;
        seen.add(name);
        actions.push({ type: "pick-discard", cardId });
      }
      return actions;
    }
    // babylon-seventh: resolve the leftover card like a normal turn.
    const leftover = state.hands[playerIndex][0];
    if (leftover === undefined) return [{ type: "skip-pending" }];
    return buildActionsForCard(state, playerIndex, leftover).map((action) => ({
      type: "play-seventh",
      action,
    }));
  }

  return [];
}

File: cpp/seven-wonders/src/rules.cpp
#include "sw/rules.hpp"

#include "sw/board.hpp"
#include "sw/edifice.hpp"
#include "sw/payment.hpp"
#include "sw/wonders.hpp"

namespace sw {

int activePlayer(const GameState& st) {
  if (st.phase == Phase::Pending && st.pendingCount > 0) return st.pendingQueue[0].player;
  return -1;
}

// Enumerate every discard / play-card / build-wonder for one hand card.
static void buildActionsForCard(const GameState& st, int player, int handIdx, bool seventh,
                                MoveBuffer& out) {
  const PlayerState& p = st.players[player];
  const Card& def = cardType(st.hands[player][handIdx]);

  // Discard is always legal.
  Move dis;
  dis.kind = MoveKind::Discard;
  dis.card = uint8_t(handIdx);
  dis.seventh = seventh;
  out.push(dis);

  // Play the card (never a duplicate name).
  if (!hasName(p, def.nameId)) {
    bool chain = false;
    for (int k = 0; k < def.numChain; k++)
      if (hasName(p, def.chainFrom[k])) chain = true;

    if (chain) {
      Move m;
      m.kind = MoveKind::PlayCard;
      m.card = uint8_t(handIdx);
      m.pay = PayKind::Chain;
      m.seventh = seventh;
      out.push(m);
    } else {
      PayResult pr = solvePayments(st, player, def.resCost, def.bankCoins);
      for (int i = 0; i < pr.count; i++) {
        Move m;
        m.kind = MoveKind::PlayCard;
        m.card = uint8_t(handIdx);
        m.pay = PayKind::Resources;
        m.payLeft = pr.splits[i].left;
        m.payRight = pr.splits[i].right;
        m.seventh = seventh;
        out.push(m);
      }
      if (hasBuiltStageEffect(p, EffectKind::FreeBuildPerAge) && !p.freeBuildUsedThisAge) {
        Move m;
        m.kind = MoveKind::PlayCard;
        m.card = uint8_t(handIdx);
        m.pay = PayKind::FreeBuild;
        m.seventh = seventh;
        out.push(m);
      }
    }
  }

  // Bury the card to build the next wonder stage (never chain/free-build).
  const WonderSide& side = wonderSide(p.wonderId, p.side);
  if (p.stagesBuilt < side.numStages) {
    const Stage& stage = side.stages[p.stagesBuilt];
    PayResult pr = solvePayments(st, player, stage.resCost, stage.bankCoins);

    // Edifice participation: this age's project, if open and not yet joined.
    const EdSlot* ed = st.edificeCount > 0 ? &st.edifices[st.age - 1] : nullptr;
    bool canParticipate = false;
    int participationCost = 0;
    if (ed && ed->status == EdStatus::Project) {
      canParticipate = true;
      for (int i = 0; i < ed->numParticipants; i++)
        if (ed->participants[i] == player) canParticipate = false;
      if (canParticipate) participationCost = EDIFICES[ed->edificeId].cost;
    }
    int stageCoinCost = stage.bankCoins;

    for (int i = 0; i < pr.count; i++) {
      Move m;
      m.kind = MoveKind::BuildWonder;
      m.card = uint8_t(handIdx);
      m.pay = PayKind::Resources;
      m.payLeft = pr.splits[i].left;
      m.payRight = pr.splits[i].right;
      m.seventh = seventh;
      out.push(m);
      int coinsLeft = p.coins - pr.splits[i].left - pr.splits[i].right - stageCoinCost;
      if (canParticipate && coinsLeft >= participationCost) {
        Move mp = m;
        mp.participate = true;
        out.push(mp);
      }
    }
  }
}

void legalActions(const GameState& st, int player, MoveBuffer& out) {
  out.clear();

  if (st.phase == Phase::Selecting) {
    if (st.hasSelection[player]) return;
    for (int h = 0; h < st.handCount[player]; h++)
      buildActionsForCard(st, player, h, false, out);
    return;
  }

  if (st.phase == Phase::Pending && st.pendingCount > 0 && st.pendingQueue[0].player == player) {
    const PendingItem& pend = st.pendingQueue[0];
    if (pend.kind == PendingKind::Halikarnassos) {
      Move skip;
      skip.kind = MoveKind::SkipPending;
      out.push(skip);
      const PlayerState& p = st.players[player];
      uint64_t seen[2] = {0, 0};
      for (int d = 0; d < st.discardCount; d++) {
        int nameId = cardType(st.discard[d]).nameId;
        if (hasName(p, nameId)) continue;
        if (seen[nameId >> 6] & (uint64_t(1) << (nameId & 63))) continue;
        seen[nameId >> 6] |= (uint64_t(1) << (nameId & 63));
        Move m;
        m.kind = MoveKind::PickDiscard;
        m.card = uint8_t(d);
        out.push(m);
      }
    } else {  // babylon-seventh: resolve the leftover 7th card like a normal turn.
      if (st.handCount[player] == 0) {
        Move skip;
        skip.kind = MoveKind::SkipPending;
        out.push(skip);
        return;
      }
      buildActionsForCard(st, player, 0, true, out);
    }
  }
}

bool isLegal(const GameState& st, int player, const Move& m) {
  MoveBuffer buf;
  legalActions(st, player, buf);
  for (int i = 0; i < buf.count; i++)
    if (buf.moves[i] == m) return true;
  return false;
}

}  // namespace sw
